### PR TITLE
Use map to speed up SpreadConstraintPair validation

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In validateTopologySpreadConstraints, currently spreadConstraintPair's are stored in an array which is iterated in ValidateSpreadConstraintPair.

This PR uses map of spreadConstraintPair to speed up validation.

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by tedyu** [comment]
cc @Huang-Wei

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -5576,7 +5576,7 @@ type spreadConstraintPair struct {
 func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstraint, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	var existingConstraintPairs []spreadConstraintPair
+	existingConstraintPairs := map[spreadConstraintPair]bool{}
```
Comment (draveness) [inline pkg/apis/core/validation/validation.go]: we could use `struct{}` instead of bool here

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -5596,7 +5596,7 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 		if err := ValidateSpreadConstraintPair(subFldPath.Child("{topologyKey, whenUnsatisfiable}"), pair, existingConstraintPairs); err != nil {
 			allErrs = append(allErrs, err)
 		} else {
-			existingConstraintPairs = append(existingConstraintPairs, pair)
+			existingConstraintPairs[pair] = true
```
Comment (draveness) [inline pkg/apis/core/validation/validation.go]: s/true/struct{}{}

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -5628,12 +5628,9 @@ func ValidateWhenUnsatisfiable(fldPath *field.Path, action core.UnsatisfiableCon
...
 
 // ValidateSpreadConstraintPair tests that if `pair` exists in `existingConstraintPairs`.
-func ValidateSpreadConstraintPair(fldPath *field.Path, pair spreadConstraintPair, existingConstraintPairs []spreadConstraintPair) *field.Error {
-	for _, existingPair := range existingConstraintPairs {
-		if pair.topologyKey == existingPair.topologyKey &&
-			pair.whenUnsatisfiable == existingPair.whenUnsatisfiable {
-			return field.Duplicate(fldPath, pair)
-		}
+func ValidateSpreadConstraintPair(fldPath *field.Path, pair spreadConstraintPair, existingConstraintPairs map[spreadConstraintPair]bool) *field.Error {
+	if existingConstraintPairs[pair] {
```
Comment (draveness) [inline pkg/apis/core/validation/validation.go]:
```
_, ok := existingConstraintPairs[pair]
if ok {
  //
}
```

**Comment by Huang-Wei** [comment]
Map is overkill here... As we don't expect users to have a lot of constraints.

I created #80395 to optimize the logic.

**Pull request closed without merging**